src/Streamly/Internal/Data/Array/Stream/Foreign.hs

```diff
@@ -480,6 +553,105 @@ parseD (PRD.Parser pstep initial extract) stream@(D.Stream step state) = do
+            PR.Partial n pst1 -> do
+                assert (n <= Prelude.length (x:getList backBuf)) (return ())
+                let src0 = Prelude.take n (x:getList backBuf)
+                    arr0 = A.fromListN n (Prelude.reverse src0)
+                    !(Ptr curAddr) = next
+                    arr1 = Array (ForeignPtr curAddr contents) (Ptr end)
+                    src = arr0 <> arr1
+                let !(Array (ForeignPtr start cont1) (Ptr end1)) = src
+                gobuf SPEC s (List []) (ForeignPtr end1 cont1) (Ptr start) pst1
+            PR.Continue 0 pst1 ->
+                gobuf SPEC s (List (x:getList backBuf)) fp next pst1
+            PR.Continue n pst1 -> do
+                assert (n <= Prelude.length (x:getList backBuf)) (return ())
+                let (src0, buf1) = splitAt n (x:getList backBuf)
+                    arr0 = A.fromListN n (Prelude.reverse src0)
+                    !(Ptr curAddr) = next
+                    arr1 = Array (ForeignPtr curAddr contents) (Ptr end)
+                    src = arr0 <> arr1
+                let !(Array (ForeignPtr start cont1) (Ptr end1)) = src
+                gobuf SPEC s (List buf1) (ForeignPtr end1 cont1) (Ptr start) pst1
+            PR.Done 0 b -> do
+                let !(Ptr curAddr) = next
+                    arr = Array (ForeignPtr curAddr contents) (Ptr end)
+                return (b, D.cons arr (D.Stream step s))
+            PR.Done n b -> do
+                assert (n <= Prelude.length (x:getList backBuf)) (return ())
+                let src0 = Prelude.take n (x:getList backBuf)
+                    -- XXX create the array in reverse instead
+                    arr0 = A.fromListN n (Prelude.reverse src0)
+                    !(Ptr curAddr) = next
+                    arr1 = Array (ForeignPtr curAddr contents) (Ptr end)
+                    -- XXX Use StreamK to avoid adding arbitrary layers of
+                    -- constructors every time.
+                    str = D.cons arr0 (D.cons arr1 (D.Stream step s))
+                return (b, str)
+            PR.Error err -> throwM $ ParseError err
+
+-- | Parse an array stream using the supplied 'Parser'.  Returns the parse
+-- result and the unconsumed stream. Throws 'ParseError' if the parse fails.
+--
+-- /Internal/
+--
+{-# INLINE_NORMAL parse #-}
+parse ::
+       (MonadIO m, MonadThrow m, Storable a)
+    => PRD.Parser m a b
+    -> SerialT m (A.Array a)
+    -> m (b, SerialT m (A.Array a))
+parse p s = fmap D.fromStreamD <$> parseD p (D.toStreamD s)
+
+-------------------------------------------------------------------------------
+-- Elimination - Running Array Folds and parsers
+-------------------------------------------------------------------------------
+
+{-# INLINE_NORMAL parseArrD #-}
+parseArrD ::
+       forall m a b. (MonadIO m, MonadThrow m, Storable a)
+    => PRD.Parser m (Array a) b
+    -> D.Stream m (Array.Array a)
+    -> m (b, D.Stream m (Array.Array a))
```

Comment:
Please remind me why we have a special parser for parsing arrays?
Shouldn't we backtrack arrays instead of backtracking elements? Did this have any use-case?

Reply:
Looks like we have similar folds as well.

Reply:
It's pre-existing code so guess there were some practical applications.

Reply:
Ok, nvm.

Reply:
We did this whole thing for streamly-lz4 and tar. We are only using foldArr_ in streamly-lz4. I don't remember if tar required parseArr as well.